File: Sources/XDecodeCore/XlogDecoder.swift
import Foundation
import P256K

public struct XlogDecoder: LogDecoder {
    public let format = LogFormat.xlog
    private let credentials: [XlogCredentials]

    public init(credentials: [XlogCredentials] = []) {
        self.credentials = credentials
    }

    public func decode(_ data: Data, sourceURL: URL) throws -> DecodedLog {
        guard !data.isEmpty else { throw DecodeError.malformed("Xlog 文件为空") }
        guard var cursor = findFrameStart(in: data) else {
            throw DecodeError.malformed("未找到有效的 Xlog 数据帧")
        }

        var output = Data()
        var diagnostics = Diagnostics()
        var previousSequence: UInt16?
        var preferredCredentialIndex: Int?

        if cursor > 0 {
            diagnostics.damagedSections += 1
            diagnostics.skippedBytes += cursor
        }

        while cursor < data.count {
            guard let frame = try? Frame(data: data, offset: cursor) else {
                diagnostics.damagedSections += 1
                if let recovered = findFrameStart(in: data, startingAt: cursor + 1) {
                    diagnostics.skippedBytes += recovered - cursor
                    cursor = recovered
                    continue
                }
                diagnostics.skippedBytes += data.count - cursor
                break
            }

            diagnostics.totalFrames += 1
            if let previousSequence,
               frame.sequence > 1,
               Int(frame.sequence) != Int(previousSequence) + 1 {
                diagnostics.sequenceGaps += 1
            }
            if frame.sequence != 0 { previousSequence = frame.sequence }

            do {
                let decoded = try decodePayload(frame, preferredCredentialIndex: &preferredCredentialIndex)
                output.append(decoded)
                diagnostics.successfulFrames += 1
            } catch PayloadError.missingCredentials {
                diagnostics.failedFrames += 1
                diagnostics.missingKeyFrames += 1
            } catch PayloadError.noCredentialSucceeded {
                diagnostics.failedFrames += 1
                diagnostics.rejectedKeyFrames += 1
            } catch {
                diagnostics.failedFrames += 1
                diagnostics.invalidPayloadFrames += 1
            }
            cursor = frame.nextOffset
        }

        guard diagnostics.successfulFrames > 0, !output.isEmpty else {
            if diagnostics.missingKeyFrames > 0 {
                throw DecodeError.decodingFailed(
                    "\(diagnostics.failureSummary)；缺少匹配的 Xlog secp256k1 私钥"
                )
            }
            if diagnostics.rejectedKeyFrames > 0 {
                throw DecodeError.decodingFailed(
                    "\(diagnostics.failureSummary)；Xlog 私钥不匹配或加密数据帧损坏"
                )
            }
            throw DecodeError.decodingFailed(
                "\(diagnostics.failureSummary)；没有可输出的日志内容"
            )
        }

        guard diagnostics.isComplete else {
            return .partial(output, diagnostic: diagnostics.message)
        }
        return .complete(output, diagnostic: diagnostics.completeMessage)
    }

    private func decodePayload(
        _ frame: Frame,
        preferredCredentialIndex: inout Int?
    ) throws -> Data {
        guard frame.magic.requiresECDH else {
            return try decompress(frame.payload, magic: frame.magic)
        }
        guard !credentials.isEmpty else { throw PayloadError.missingCredentials }

        var candidateIndices = Array(credentials.indices)
        if let preferredCredentialIndex,
           let index = candidateIndices.firstIndex(of: preferredCredentialIndex) {
            candidateIndices.remove(at: index)
            candidateIndices.insert(preferredCredentialIndex, at: 0)
        }

        for index in candidateIndices {
            do {
                let teaKey = try deriveTEAKey(
                    publicKey: frame.publicKey,
                    privateKeyData: credentials[index].privateKey
                )
                let decrypted = TEA.decrypt(frame.payload, key: teaKey)
                let decoded = try decompress(decrypted, magic: frame.magic)
                preferredCredentialIndex = index
                return decoded
            } catch {
                continue
            }
        }
        throw PayloadError.noCredentialSucceeded
    }

    private func decompress(_ payload: Data, magic: XlogMagic) throws -> Data {
        switch magic {
        case .compress, .compressNoCrypt, .compressECDH:
            return try CompressionUtilities.inflateRaw(payload)
        case .compressChunked:
            return try CompressionUtilities.inflateRaw(try joinChunkedPayload(payload))
        case .syncZstdCrypt, .syncZstdNoCrypt, .asyncZstdCrypt, .asyncZstdNoCrypt:
            return try CompressionUtilities.zstd(payload)
        case .noCompress, .noCompressExtended, .noCompressNoCrypt:
            return payload
        }
    }

    private func joinChunkedPayload(_ data: Data) throws -> Data {
        var cursor = 0
        var joined = Data()
        while cursor < data.count {
            let length = Int(try data.uint16LE(at: cursor))
            cursor += 2
            joined.append(try data.checkedRange(cursor..<(cursor + length)))
            cursor += length
        }
        return joined
    }

    private func deriveTEAKey(publicKey: Data, privateKeyData: Data) throws -> Data {
        guard publicKey.count == 64 else {
            throw DecodeError.decryptionFailed("Xlog ECDH 公钥长度无效")
        }
        var x963 = Data([0x04])
        x963.append(publicKey)

        do {
            let privateKey = try P256K.KeyAgreement.PrivateKey(
                dataRepresentation: privateKeyData,
                format: .uncompressed
            )
            let peerKey = try P256K.KeyAgreement.PublicKey(x963Representation: x963)
            let sharedSecret = privateKey.sharedSecretFromKeyAgreement(with: peerKey, format: .uncompressed)
            let sharedPoint = sharedSecret.withUnsafeBytes { Data($0) }
            guard sharedPoint.count == 65 else {
                throw DecodeError.decryptionFailed("ECDH 共享点长度无效")
            }
            return sharedPoint.subdata(in: 1..<17)
        } catch let error as DecodeError {
            throw error
        } catch {
            throw DecodeError.decryptionFailed("ECDH：\(error.localizedDescription)")
        }
    }

    private func findFrameStart(in data: Data, startingAt start: Int = 0) -> Int? {
        guard start < data.count else { return nil }
        for offset in start..<data.count {
            if (try? Frame(data: data, offset: offset)) != nil { return offset }
        }
        return nil
    }
}

private extension XlogDecoder {
    enum PayloadError: Error {
        case missingCredentials
        case noCredentialSucceeded
    }

    struct Diagnostics {
        var totalFrames = 0
        var successfulFrames = 0
        var failedFrames = 0
        var missingKeyFrames = 0
        var rejectedKeyFrames = 0
        var invalidPayloadFrames = 0
        var damagedSections = 0
        var skippedBytes = 0
        var sequenceGaps = 0

        var isComplete: Bool {
            failedFrames == 0 && damagedSections == 0 && sequenceGaps == 0
        }

        var message: String {
            var details = ["Xlog 部分解密：成功 \(successfulFrames) 帧，失败 \(failedFrames) 帧"]
            if missingKeyFrames > 0 {
                details.append("\(missingKeyFrames) 个加密帧缺少匹配私钥")
            }
            if rejectedKeyFrames > 0 {
                details.append("\(rejectedKeyFrames) 个加密帧密钥不匹配或已损坏")
            }
            if invalidPayloadFrames > 0 {
                details.append("\(invalidPayloadFrames) 个数据帧解压失败或已损坏")
            }
            if damagedSections > 0 {
                details.append("跳过 \(damagedSections) 段损坏数据（\(skippedBytes) 字节）")
            }
            if sequenceGaps > 0 {
                details.append("检测到 \(sequenceGaps) 处日志序号缺失")
            }
            return details.joined(separator: "；")
        }

        var completeMessage: String {
            "Xlog 解密完成：成功 \(successfulFrames) 帧，失败 0 帧"
        }

        var failureSummary: String {
            "Xlog 解密失败：成功 0 帧，失败 \(failedFrames) 帧"
        }
    }
}

private enum XlogMagic: UInt8 {
    case noCompress = 0x03
    case compress = 0x04
    case compressChunked = 0x05
    case noCompressExtended = 0x06
    case compressECDH = 0x07
    case noCompressNoCrypt = 0x08
    case compressNoCrypt = 0x09
    case syncZstdCrypt = 0x0A
    case syncZstdNoCrypt = 0x0B
    case asyncZstdCrypt = 0x0C
    case asyncZstdNoCrypt = 0x0D

    var keyLength: Int {
        switch self {
        case .noCompress, .compress, .compressChunked: 4
        default: 64
        }
    }

    var requiresECDH: Bool {
        switch self {
        case .compressECDH, .syncZstdCrypt, .asyncZstdCrypt: true
        default: false
        }
    }
}

private struct Frame {
    let magic: XlogMagic
    let sequence: UInt16
    let publicKey: Data
    let payload: Data
    let nextOffset: Int

    init(data: Data, offset: Int) throws {
        guard let magic = XlogMagic(rawValue: try data.uint8(at: offset)) else {
            throw DecodeError.malformed("Xlog Magic 无效")
        }
        let headerLength = 1 + 2 + 1 + 1 + 4 + magic.keyLength
        guard offset + headerLength + 1 <= data.count else {
            throw DecodeError.malformed("Xlog Header 不完整")
        }
        let payloadLength = Int(try data.uint32LE(at: offset + 5))
        let payloadStart = offset + headerLength
        let endMarker = payloadStart + payloadLength
        guard payloadLength >= 0, endMarker < data.count, try data.uint8(at: endMarker) == 0 else {
            throw DecodeError.malformed("Xlog 数据帧长度或结束标记无效")
        }

        self.magic = magic
        self.sequence = try data.uint16LE(at: offset + 1)
        self.publicKey = try data.checkedRange((offset + 9)..<(offset + 9 + magic.keyLength))
        self.payload = try data.checkedRange(payloadStart..<endMarker)
        self.nextOffset = endMarker + 1
    }
}

private enum TEA {
    static func decrypt(_ data: Data, key: Data) -> Data {
        guard key.count >= 16 else { return data }
        let completeLength = data.count / 8 * 8
        var output = Data(capacity: data.count)
        let keyWords = (0..<4).map { keyWord(key, offset: $0 * 4) }

        var offset = 0
        while offset < completeLength {
            var v0 = keyWord(data, offset: offset)
            var v1 = keyWord(data, offset: offset + 4)
            let delta: UInt32 = 0x9E37_79B9
            var sum = delta &<< 4

            for _ in 0..<16 {
                v1 = v1 &- (((v0 &<< 4) &+ keyWords[2]) ^ (v0 &+ sum) ^ ((v0 >> 5) &+ keyWords[3]))
                v0 = v0 &- (((v1 &<< 4) &+ keyWords[0]) ^ (v1 &+ sum) ^ ((v1 >> 5) &+ keyWords[1]))
                sum = sum &- delta
            }
            appendLittleEndian(v0, to: &output)
            appendLittleEndian(v1, to: &output)
            offset += 8
        }
        if completeLength < data.count { output.append(data.suffix(from: completeLength)) }
        return output
    }

    private static func keyWord(_ data: Data, offset: Int) -> UInt32 {
        (try? data.uint32LE(at: offset)) ?? 0
    }

    private static func appendLittleEndian(_ value: UInt32, to data: inout Data) {
        data.append(UInt8(truncatingIfNeeded: value))
        data.append(UInt8(truncatingIfNeeded: value >> 8))
        data.append(UInt8(truncatingIfNeeded: value >> 16))
        data.append(UInt8(truncatingIfNeeded: value >> 24))
    }
}

File: Windows/src/XDecode.Core/XlogDecoder.cs
using System.Buffers.Binary;
using Org.BouncyCastle.Asn1.Sec;
using Org.BouncyCastle.Math;

namespace XDecode.Core;

public sealed class XlogDecoder(IEnumerable<XlogCredentials>? credentials = null) : ILogDecoder
{
    private readonly XlogCredentials[] _credentials = credentials?.ToArray() ?? [];
    public LogFormat Format => LogFormat.Xlog;

    public DecodedLog Decode(ReadOnlyMemory<byte> data, string sourcePath)
    {
        if (data.IsEmpty) throw DecodeException.Malformed("Xlog 文件为空");
        var span = data.Span;
        var cursor = FindFrameStart(span, 0);
        if (cursor < 0) throw DecodeException.Malformed("未找到有效的 Xlog 数据帧");

        using var output = new MemoryStream();
        var diagnostics = new Diagnostics();
        ushort? previousSequence = null;
        int? preferredCredentialIndex = null;
        if (cursor > 0)
        {
            diagnostics.DamagedSections++;
            diagnostics.SkippedBytes += cursor;
        }

        while (cursor < span.Length)
        {
            if (!Frame.TryRead(span, cursor, out var frame))
            {
                diagnostics.DamagedSections++;
                var recovered = FindFrameStart(span, cursor + 1);
                if (recovered >= 0)
                {
                    diagnostics.SkippedBytes += recovered - cursor;
                    cursor = recovered;
                    continue;
                }
                diagnostics.SkippedBytes += span.Length - cursor;
                break;
            }

            diagnostics.TotalFrames++;
            if (previousSequence.HasValue && frame.Sequence > 1 &&
                frame.Sequence != previousSequence.Value + 1)
                diagnostics.SequenceGaps++;
            if (frame.Sequence != 0) previousSequence = frame.Sequence;

            try
            {
                var maximum = checked(DecodeLimits.MaximumDecompressedOutputSize - (int)output.Length);
                var decoded = DecodePayload(frame, ref preferredCredentialIndex, maximum);
                DecodeLimits.ValidateOutputSize(output.Length, decoded.Length);
                output.Write(decoded);
                diagnostics.SuccessfulFrames++;
            }
            catch (PayloadException exception) when (exception.MissingCredentials)
            {
                diagnostics.FailedFrames++;
                diagnostics.MissingKeyFrames++;
            }
            catch (PayloadException)
            {
                diagnostics.FailedFrames++;
                diagnostics.RejectedKeyFrames++;
            }
            catch (DecodeException exception) when (exception.Kind == DecodeErrorKind.OutputLimitExceeded)
            {
                throw;
            }
            catch
            {
                diagnostics.FailedFrames++;
                diagnostics.InvalidPayloadFrames++;
            }
            cursor = frame.NextOffset;
        }

        var bytes = output.ToArray();
        if (diagnostics.SuccessfulFrames == 0 || bytes.Length == 0)
        {
            if (diagnostics.MissingKeyFrames > 0)
                throw DecodeException.DecodingFailed($"{diagnostics.FailureSummary}；缺少匹配的 Xlog secp256k1 私钥");
            if (diagnostics.RejectedKeyFrames > 0)
                throw DecodeException.DecodingFailed($"{diagnostics.FailureSummary}；Xlog 私钥不匹配或加密数据帧损坏");
            throw DecodeException.DecodingFailed($"{diagnostics.FailureSummary}；没有可输出的日志内容");
        }
        return diagnostics.IsComplete
            ? DecodedLog.Complete(bytes, diagnostics.CompleteMessage)
            : DecodedLog.Partial(bytes, diagnostics.Message);
    }

    private byte[] DecodePayload(Frame frame, ref int? preferredCredentialIndex, int maximumOutputSize)
    {
        if (!Info(frame.Magic).RequiresEcdh)
            return Decompress(frame.Payload, frame.Magic, maximumOutputSize);
        if (_credentials.Length == 0) throw PayloadException.Missing();

        var indices = Enumerable.Range(0, _credentials.Length).ToList();
        if (preferredCredentialIndex is { } preferred && indices.Remove(preferred))
            indices.Insert(0, preferred);
        foreach (var index in indices)
        {
            try
            {
                var key = DeriveTeaKey(frame.PublicKey, _credentials[index].PrivateKey);
                var decrypted = Tea.Decrypt(frame.Payload, key);
                var decoded = Decompress(decrypted, frame.Magic, maximumOutputSize);
                preferredCredentialIndex = index;
                return decoded;
            }
            catch (DecodeException exception) when (exception.Kind == DecodeErrorKind.OutputLimitExceeded)
            {
                throw;
            }
            catch { }
        }
        throw PayloadException.Rejected();
    }

    private static byte[] Decompress(ReadOnlySpan<byte> payload, XlogMagic magic, int maximumOutputSize) =>
        magic switch
        {
            XlogMagic.Compress or XlogMagic.CompressNoCrypt or XlogMagic.CompressEcdh =>
                CompressionUtilities.InflateRaw(payload, maximumOutputSize),
            XlogMagic.CompressChunked =>
                CompressionUtilities.InflateRaw(JoinChunkedPayload(payload), maximumOutputSize),
            XlogMagic.SyncZstdCrypt or XlogMagic.SyncZstdNoCrypt or
                XlogMagic.AsyncZstdCrypt or XlogMagic.AsyncZstdNoCrypt =>
                CompressionUtilities.Zstd(payload, maximumOutputSize),
            _ => CopyBounded(payload, maximumOutputSize)
        };

    private static byte[] CopyBounded(ReadOnlySpan<byte> payload, int maximumOutputSize)
    {
        DecodeLimits.ValidateOutputSize(0, payload.Length, maximumOutputSize);
        return payload.ToArray();
    }

    private static byte[] JoinChunkedPayload(ReadOnlySpan<byte> payload)
    {
        using var output = new MemoryStream();
        var cursor = 0;
        while (cursor < payload.Length)
        {
            var length = payload.UInt16LE(cursor);
            cursor += 2;
            output.Write(payload.CheckedRange(cursor, length));
            cursor += length;
        }
        return output.ToArray();
    }

    private static byte[] DeriveTeaKey(ReadOnlySpan<byte> publicKey, ReadOnlySpan<byte> privateKey)
    {
        if (publicKey.Length != 64) throw DecodeException.DecryptionFailed("Xlog ECDH 公钥长度无效");
        try
        {
            var domain = SecNamedCurves.GetByName("secp256k1");
            var encoded = new byte[65];
            encoded[0] = 0x04;
            publicKey.CopyTo(encoded.AsSpan(1));
            var point = domain.Curve.DecodePoint(encoded);
            var scalar = new BigInteger(1, privateKey.ToArray());
            var shared = point.Multiply(scalar).Normalize().GetEncoded(false);
            if (shared.Length != 65) throw DecodeException.DecryptionFailed("ECDH 共享点长度无效");
            return shared[1..17];
        }
        catch (DecodeException) { throw; }
        catch (Exception exception) { throw DecodeException.DecryptionFailed($"ECDH：{exception.Message}", exception); }
    }

    private static int FindFrameStart(ReadOnlySpan<byte> data, int start)
    {
        for (var offset = Math.Max(start, 0); offset < data.Length; offset++)
            if (Frame.TryRead(data, offset, out _)) return offset;
        return -1;
    }

    private sealed class PayloadException(bool missing) : Exception
    {
        public bool MissingCredentials { get; } = missing;
        public static PayloadException Missing() => new(true);
        public static PayloadException Rejected() => new(false);
    }

    private sealed class Diagnostics
    {
        public int TotalFrames, SuccessfulFrames, FailedFrames, MissingKeyFrames, RejectedKeyFrames;
        public int InvalidPayloadFrames, DamagedSections, SkippedBytes, SequenceGaps;
        public bool IsComplete => FailedFrames == 0 && DamagedSections == 0 && SequenceGaps == 0;
        public string CompleteMessage => $"Xlog 解密完成：成功 {SuccessfulFrames} 帧，失败 0 帧";
        public string FailureSummary => $"Xlog 解密失败：成功 0 帧，失败 {FailedFrames} 帧";
        public string Message
        {
            get
            {
                var details = new List<string> { $"Xlog 部分解密：成功 {SuccessfulFrames} 帧，失败 {FailedFrames} 帧" };
                if (MissingKeyFrames > 0) details.Add($"{MissingKeyFrames} 个加密帧缺少匹配私钥");
                if (RejectedKeyFrames > 0) details.Add($"{RejectedKeyFrames} 个加密帧密钥不匹配或已损坏");
                if (InvalidPayloadFrames > 0) details.Add($"{InvalidPayloadFrames} 个数据帧解压失败或已损坏");
                if (DamagedSections > 0) details.Add($"跳过 {DamagedSections} 段损坏数据（{SkippedBytes} 字节）");
                if (SequenceGaps > 0) details.Add($"检测到 {SequenceGaps} 处日志序号缺失");
                return string.Join('；', details);
            }
        }
    }

    private enum XlogMagic : byte
    {
        NoCompress = 0x03, Compress = 0x04, CompressChunked = 0x05,
        NoCompressExtended = 0x06, CompressEcdh = 0x07, NoCompressNoCrypt = 0x08,
        CompressNoCrypt = 0x09, SyncZstdCrypt = 0x0a, SyncZstdNoCrypt = 0x0b,
        AsyncZstdCrypt = 0x0c, AsyncZstdNoCrypt = 0x0d
    }

    private readonly record struct MagicInfo(int KeyLength, bool RequiresEcdh);

    private static MagicInfo Info(XlogMagic magic) => magic switch
    {
        XlogMagic.NoCompress or XlogMagic.Compress or XlogMagic.CompressChunked => new(4, false),
        XlogMagic.CompressEcdh or XlogMagic.SyncZstdCrypt or XlogMagic.AsyncZstdCrypt => new(64, true),
        _ => new(64, false)
    };

    private readonly record struct Frame(
        XlogMagic Magic, ushort Sequence, byte[] PublicKey, byte[] Payload, int NextOffset)
    {
        public static bool TryRead(ReadOnlySpan<byte> data, int offset, out Frame frame)
        {
            frame = default;
            try
            {
                var rawMagic = data.UInt8(offset);
                if (rawMagic is < 0x03 or > 0x0d) return false;
                var magic = (XlogMagic)rawMagic;
                var info = Info(magic);
                var headerLength = 1 + 2 + 1 + 1 + 4 + info.KeyLength;
                if (offset < 0 || offset > data.Length - headerLength - 1) return false;
                var payloadLength = checked((int)data.UInt32LE(offset + 5));
                var payloadStart = checked(offset + headerLength);
                var endMarker = checked(payloadStart + payloadLength);
                if (endMarker >= data.Length || data[endMarker] != 0) return false;
                frame = new(
                    magic,
                    data.UInt16LE(offset + 1),
                    data.CheckedRange(offset + 9, info.KeyLength).ToArray(),
                    data.CheckedRange(payloadStart, payloadLength).ToArray(),
                    endMarker + 1);
                return true;
            }
            catch { return false; }
        }
    }

    private static class Tea
    {
        public static byte[] Decrypt(ReadOnlySpan<byte> data, ReadOnlySpan<byte> key)
        {
            if (key.Length < 16) return data.ToArray();
            var output = new byte[data.Length];
            var keyWords = new uint[4];
            for (var index = 0; index < 4; index++)
                keyWords[index] = BinaryPrimitives.ReadUInt32LittleEndian(key.Slice(index * 4, 4));
            var completeLength = data.Length / 8 * 8;
            for (var offset = 0; offset < completeLength; offset += 8)
            {
                var v0 = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
                var v1 = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset + 4, 4));
                const uint delta = 0x9e3779b9;
                var sum = unchecked(delta << 4);
                for (var round = 0; round < 16; round++)
                {
                    v1 = unchecked(v1 - (((v0 << 4) + keyWords[2]) ^ (v0 + sum) ^ ((v0 >> 5) + keyWords[3])));
                    v0 = unchecked(v0 - (((v1 << 4) + keyWords[0]) ^ (v1 + sum) ^ ((v1 >> 5) + keyWords[1])));
                    sum = unchecked(sum - delta);
                }
                BinaryPrimitives.WriteUInt32LittleEndian(output.AsSpan(offset, 4), v0);
                BinaryPrimitives.WriteUInt32LittleEndian(output.AsSpan(offset + 4, 4), v1);
            }
            data[completeLength..].CopyTo(output.AsSpan(completeLength));
            return output;
        }
    }
}
